Tutorials for integrating external party libaries into holoscan SDK (#281)
nvidia-holoscan/holohub: tutorials/integrate_3rd_libs_into_pipeline/README.md

> # Best Practices to integrate external libraries into Holoscan pipelines

Reviewer: I like the rewording from "third party" to "external", as many of the libraries discussed here are developed in part by NVIDIA.

Could you please rename the project folder to match the new title? I.e., rename from `integrate_3rd_libs_into_pipeline` to `integrate_external_libs_into_pipeline`.

        cp_array = img_as_float(cp_array)
        cp_res=cu_exposure.equalize_adapthist(cp_array)
        cp_array = img_as_ubyte(cp_res)

        # Use cupy array memory as holoscan tensor for following operations
        hs_output = hs.as_tensor(cp_array)
        out_message = Entity(context)
        out_message.add(hs_output)
        op_output.emit(out_message, "output_tensor")

```

## Integrate CV-CUDA library
[CV-CUDA](https://github.com/CVCUDA/CV-CUDA) is an open-source, graphics processing unit (GPU)-accelerated library for cloud-scale image processing and computer vision developed jointly by NVIDIA and the ByteDance Applied Machine Learning teams. CV-CUDA helps developers build highly efficient pre- and post-processing pipelines that can improve throughput by more than 10x while lowering cloud computing costs.

See the supported CV-CUDA Operators in the [CV-CUDA developer guide](https://github.com/CVCUDA/CV-CUDA/blob/main/DEVELOPER_GUIDE.md)

### Installation
Follow the [CV-CUDA documentation](https://cvcuda.github.io/installation.html) to install the CV-CUDA library.

Requirement: CV-CUDA >= 0.2.1 (From which version DLPack interop is supported)

### Sample code 

> CV-CUDA implemented with DLPack standards. So, CV-CUDA tensor can directly access Holocan Tensor. 

Reviewer: ```suggestion
CV-CUDA is implemented with DLPack standards. So, CV-CUDA tensor can directly access Holocan Tensor. 
```

       
        cvcuda_resize_tensor = cvcuda.resize(
                    cvcuda_input_tensor,
                    (
                        640,
                        640,
                        3,
                    ),
                    cvcuda.Interp.LINEAR,
                )
       
        buffer = cvcuda_resize_tensor.cuda()

        hs_output = hs.as_tensor(buffer)
        out_message = Entity(context)
        out_message.add(hs_output)
        op_output.emit(out_message, "output_tensor")

```

## Integrate OpenCV with CUDA Module
[OpenCV](https://opencv.org/) (Open Source Computer Vision Library) is a comprehensive open-source library that contains over 2500 algorithms covering Image & Video Manipulation, Object and Face Detection, OpenCV Deep Learning Module and much more. 

[OpenCV also supports GPU acceleration](https://docs.opencv.org/4.8.0/d2/dbc/cuda_intro.html), includes a CUDA module which is a set of classes and functions to utilize CUDA computational capabilities. It is implemented using NVIDIA CUDA Runtime API and provides utility functions, low-level vision primitives, and high-level algorithms.

### Installation

> Pre-Requisition:

Reviewer: ```suggestion
Prerequisites:
```

> - OpenCV >= 4.8.0 (From which version, OpenCV GpuMat supports initialization with GPU Memory pointer)
> Install OpenCV with its CUDA module following the guide in [opencv/opencv_contrib](https://github.com/opencv/opencv_contrib/tree/4.x) 

Reviewer: ```suggestion
- OpenCV >= 4.8.0 (From which version, OpenCV GpuMat supports initialization with GPU Memory pointer)

Install OpenCV with its CUDA module following the guide in [opencv/opencv_contrib](https://github.com/opencv/opencv_contrib/tree/4.x) 
```

Reviewer: The markdown document does not get formatted correctly without this extra line.

> # Best Practice to integrate 3rd party libaries into pipeline based on Holoscan SDK

Reviewer 2: ```suggestion
# Best Practices to integrate external libraries into Holoscan pipelines
```

<div align="center">
<img src="./images/typical_pipeline_holoscan.png" style="border: 2px solid black;">
</div>

The datatype in Holoscan SDK is defined as [Tensor](https://docs.nvidia.com/holoscan/sdk-user-guide/generated/classholoscan_1_1tensor.html) which is a multi-dimensional array of elements of a single data type. The Tensor class is a wrapper around the [DLManagedTensorCtx](https://docs.nvidia.com/holoscan/sdk-user-guide/generated/structholoscan_1_1dlmanagedtensorctx.html#structholoscan_1_1DLManagedTensorCtx) struct that holds the DLManagedTensor object. It also supports both DLPack and NumPy’s array interface (__array_interface__ and __cuda_array_interface__) so that it can be used with other Python libraries such as [CuPy](https://docs.cupy.dev/en/stable/user_guide/interoperability.html), [PyTorch](https://github.com/pytorch/pytorch/issues/15601), [JAX](https://github.com/google/jax/issues/1100#issuecomment-580773098), [TensorFlow](https://github.com/tensorflow/community/pull/180), and [Numba](https://numba.readthedocs.io/en/stable/cuda/cuda_array_interface.html).

> In this tutorials, we will show how to integrate below libaries into Holoscan applications in Python language:

Reviewer 2: ```suggestion
In this tutorial, we will show how to integrate the libraries below into Holoscan applications in Python:
```

> - [Integrate RAPIDS **cuCIM** libary](#integrate-rapids-cucim-libary)

Reviewer 2: There are 16 match for libary instead of library (and libaries instead of libraries) that could be addressed

Reviewer 3: Please check the [check-lint results](https://github.com/nvidia-holoscan/holohub/actions/runs/8747200842/job/24035222859?pr=281)

- [Integrate RAPIDS **CV-CUDA** libary](#integrate-cv-cuda-libary)
- [Integrate **OpenCV with CUDA Module**](#integrate-opencv-with-cuda-module)

## Integrate RAPIDS cuCIM libary
[RAPIDS cuCIM](https://github.com/rapidsai/cucim) (Compute Unified Device Architecture Clara IMage) is an open-source, accelerated computer vision and image processing software library for multidimensional images used in biomedical, geospatial, material and life science, and remote sensing use cases.

See the supported Operators in [cuCIM documentation](https://docs.rapids.ai/api/cucim/stable/).

According to the documentation, cuCIM is interoperable with CuPy. So we can using CuPy array initialized from Holoscan Tensor directly in cuCIM operators.

Per the documentation, cuCIM offers interoperability with CuPy. So we can initialize CuPy arrays directly from Holoscan Tensors which can be used in cuCIM operators for processing without memory transfer between host and device. 
### Installation

> Following the [guidance](https://github.com/rapidsai/cucim?tab=readme-ov-file#install-cucim) to install the RAPIDS cuCIM libary.

Reviewer: ```suggestion
Follow the [cuCIM documentation](https://github.com/rapidsai/cucim?tab=readme-ov-file#install-cucim) to install the RAPIDS cuCIM libary.
```

def CustomizedcuCIMOperator(Operator):
    ### Other implementation of __init__, setup()... etc. 

    def compute(self, op_input, op_output, context):
        message = op_input.receive("input_tensor")
        input_tensor = message.get()
        # Directly use Holoscan tensor to initialize cupy array
        cp_array = cp.asarray(input_tensor)

        cp_array = img_as_float(cp_array)
        cp_res=cu_exposure.equalize_adapthist(cp_array)
        cp_array = img_as_ubyte(cp_res)

        # Use cupy array memory as holoscan tensor for following operations
        hs_output = hs.as_tensor(cp_array)
        out_message = Entity(context)
        out_message.add(hs_output)
        op_output.emit(out_message, "output_tensor")

```

## Integrate CV-CUDA libary
[CV-CUDA](https://github.com/CVCUDA/CV-CUDA) is an open-source, graphics processing unit (GPU)-accelerated library for cloud-scale image processing and computer vision developed jointly by NVIDIA and the ByteDance Applied Machine Learning teams. CV-CUDA helps developers build highly efficient pre- and post-processing pipelines that can improve throughput by more than 10x while lowering cloud computing costs.

> See the supported Operators in [CV-CUDA developer guide](https://github.com/CVCUDA/CV-CUDA/blob/main/DEVELOPER_GUIDE.md)

Reviewer: ```suggestion
See the supported CV-CUDA Operators in the [CV-CUDA developer guide](https://github.com/CVCUDA/CV-CUDA/blob/main/DEVELOPER_GUIDE.md)
```

### Installation

> Following the [guidance](https://cvcuda.github.io/installation.html) to install the CV-CUDA libary.

Reviewer: ```suggestion
Follow the [CV-CUDA documentation](https://cvcuda.github.io/installation.html) to install the CV-CUDA libary.
```

Requirement: CV-CUDA >= 0.2.1 (From which version DLPack interop is supported)

### Sample code 
CV-CUDA implemented with DLPack standards. So, CV-CUDA tensor can directly access Holocan Tensor. 

> Detial usage with CV-CUDA in Holoscan application can refer this [sample application](https://github.com/nvidia-holoscan/holohub/tree/main/applications/cvcuda_basic). 

Reviewer: ```suggestion
Refer to the [Holoscan CV-CUDA sample application](https://github.com/nvidia-holoscan/holohub/tree/main/applications/cvcuda_basic) for an example of how to use CV-CUDA with Holoscan SDK.
```

       
        cvcuda_resize_tensor = cvcuda.resize(
                    cvcuda_input_tensor,
                    (
                        640,
                        640,
                        3,
                    ),
                    cvcuda.Interp.LINEAR,
                )
       
        buffer = cvcuda_resize_tensor.cuda()

        hs_output = hs.as_tensor(buffer)
        out_message = Entity(context)
        out_message.add(hs_output)
        op_output.emit(out_message, "output_tensor")

```

## Integrate OpenCV with CUDA Module
[OpenCV](https://opencv.org/) (Open Source Computer Vision Library) is a comprehensive open-source library that contains over 2500 algorithms covering Image & Video Manipulation, Object and Face Detection, OpenCV Deep Learning Module and much more. 

[OpenCV also supports GPU acceleration](https://docs.opencv.org/4.8.0/d2/dbc/cuda_intro.html), includes a CUDA module which is a set of classes and functions to utilize CUDA computational capabilities. It is implemented using NVIDIA CUDA Runtime API and provides utility functions, low-level vision primitives, and high-level algorithms.

### Installation
Pre-Requisition:

> - OpenCV >= 4.8.0 (From which version, OpenCV GpuMat support to be initialized with GPU Memory pointer)

Reviewer: ```suggestion
- OpenCV >= 4.8.0 (From which version, OpenCV GpuMat supports initialization with GPU Memory pointer)
```

> Install OpenCV with CUDA module following the guide in [opencv/opencv_contrib](https://github.com/opencv/opencv_contrib/tree/4.x) 

Reviewer: ```suggestion
Install OpenCV with its CUDA module following the guide in [opencv/opencv_contrib](https://github.com/opencv/opencv_contrib/tree/4.x) 
```

> Also recommend to refer this [Dockerfile](https://github.com/nvidia-holoscan/holohub/blob/main/applications/endoscopy_depth_estimation/Dockerfile) to build an image with Holoscan SDK and OpenCV CUDA. 

Reviewer: ```suggestion
We also recommend referring to the [Holoscan Endoscopy Depth Estimation application container](https://github.com/nvidia-holoscan/holohub/blob/main/applications/endoscopy_depth_estimation/Dockerfile) as an example of how to build an image with Holoscan SDK and OpenCV CUDA. 
```

### Sample code
The datatype of OpenCV is GpuMat which implements neither the __cuda_array_interface__ nor the standard DLPack. To achieve the end-to-end GPU accelerated pipeline / application, we need to implement 2 functions to convert the GpuMat to CuPy array which can be accessed directly with Holoscan Tensor and vice versa. 

> Detial usage with OpenCV Operator in Holoscan application can refer this [sample application](https://github.com/nvidia-holoscan/holohub/tree/main/applications/endoscopy_depth_estimation). 

Reviewer: ```suggestion
Refer to the [Holoscan Endoscopy Depth Estimation sample application](https://github.com/nvidia-holoscan/holohub/tree/main/applications/cvcuda_basic) for an example of how to use the OpenCV operator with Holoscan SDK.
```

1. Conversion from GpuMat to CuPy Array

The GpuMat object of OpenCV Python bindings provides a cudaPtr method which can be used to access the GPU memory address of a GpuMat object. This memory pointer can be utilized to initialize a CuPy array directly, allowing for efficient data handling by avoiding unnecessary data transfers between the host and device. 

> Refer to the function below, which is used to create a CuPy array from a GpuMat. For more details, see the source code in the [holohub/applications/endoscopy_depth_estimation-gpumat_to_cupy](https://github.com/nvidia-holoscan/holohub/blob/main/applications/endoscopy_depth_estimation/endoscopy_depth_estimation.py#L52). 

Reviewer: ```suggestion
Refer to the function below, which is used to create a CuPy array from a GpuMat. For more details, see the source code in [holohub/applications/endoscopy_depth_estimation-gpumat_to_cupy](https://github.com/nvidia-holoscan/holohub/blob/main/applications/endoscopy_depth_estimation/endoscopy_depth_estimation.py#L52). 
```

```
import cv2
import cupy as cp

def gpumat_to_cupy(gpu_mat: cv2.cuda.GpuMat) -> cp.ndarray:
    w, h = gpu_mat.size()
    size_in_bytes = gpu_mat.step * w
    shapes = (h, w, gpu_mat.channels())
    assert gpu_mat.channels() <=3, "Unsupported GpuMat channels"

    dtype = None
    if gpu_mat.type() in [cv2.CV_8U,cv2.CV_8UC1,cv2.CV_8UC2,cv2.CV_8UC3]:
        dtype = cp.uint8
    elif gpu_mat.type() == cv2.CV_8S:
        dtype = cp.int8
    elif gpu_mat.type() == cv2.CV_16U:
        dtype = cp.uint16
    elif gpu_mat.type() == cv2.CV_16S:
        dtype = cp.int16
    elif gpu_mat.type() == cv2.CV_32S:
        dtype = cp.int32
    elif gpu_mat.type() == cv2.CV_32F:
        dtype = cp.float32
    elif gpu_mat.type() == cv2.CV_64F:
        dtype = cp.float64 

    assert dtype is not None, "Unsupported GpuMat type"
    
    mem = cp.cuda.UnownedMemory(gpu_mat.cudaPtr(), size_in_bytes, owner=gpu_mat)
    memptr = cp.cuda.MemoryPointer(mem, offset=0)
    cp_out = cp.ndarray(
        shapes,
        dtype=dtype,
        memptr=memptr,
        strides=(gpu_mat.step, gpu_mat.elemSize(), gpu_mat.elemSize1()),
    )
    return cp_out

```

> Note: In this function we used [UnownedMemory](https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.UnownedMemory.html#cupy.cuda.UnownedMemory) API to create the CuPy array. Since in some cases, the OpenCV operators will create a new device memory which needs to be handled by CuPy and the lifetime is not limited to one operator but the whole pipeline. In this case, the CuPy array initiated from the GpuMat shall know the owner and keep the reference to the object, check more detail on [CuPy interoperability](https://docs.cupy.dev/en/stable/user_guide/interoperability.html#device-memory-pointers).

Reviewer: ```suggestion
Note: In this function we used the [UnownedMemory](https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.UnownedMemory.html#cupy.cuda.UnownedMemory) API to create the CuPy array. In some cases, the OpenCV operators will allocate new device memory which needs to be handled by CuPy and the lifetime is not limited to one operator but the whole pipeline. In this case, the CuPy array initiated from the GpuMat shall know the owner and keep the reference to the object. Check the CuPy documentation for more detail on [CuPy interoperability](https://docs.cupy.dev/en/stable/user_guide/interoperability.html#device-memory-pointers).
```

2. Conversion from Holoscan Tensor to GpuMat via CuPy array

With the release of OpenCV 4.8, the Python bindings for OpenCV now support the initialization of GpuMat objects directly from GPU memory pointers. This capability facilitates more efficient data handling and processing by allowing direct interaction with GPU-resident data, bypassing the need for data transfer between host and device memory. 

> Within pipeline applications based Holoscan SDK, the GPU Memory pointer can be obtained through the __cuda_array_interface__ interface provided by CuPy arrays. 

Reviewer: ```suggestion
Within pipeline applications based on Holoscan SDK, the GPU Memory pointer can be obtained through the `__cuda_array_interface__` interface provided by CuPy arrays. 
```
